### PR TITLE
Feature/stat cal tackles statistics

### PR DESCRIPTION
**Pull request by kevin-newland**
This pull request has a purpose of adding features:
The tests that are added: most_tackles, and fewest_tackles
The methods for those tests are:  most_tackles, and fewest_tackles

### CODE REVIEW

File: lib/stat_calculator.rb [hunk 1]
```diff
@@ -171,4 +171,68 @@ def worst_offense
...
+    end
+    
+    team_with_most_tackles = nil
+    most_tackles = 0
+    team_tackles.each do |team_id, tackles|
+      if tackles > most_tackles
+        most_tackles = tackles
+        team_with_most_tackles = team_id
+      end
+    end
```
Comment (DustinPeukert) [inline lib/stat_calculator.rb:198]: I could be wrong but I think the block above could be accomplished a bit more clearly with a max_by enumerable!

File: lib/stat_calculator.rb [hunk 2]
```diff
@@ -171,4 +171,68 @@ def worst_offense
...
+    team_tackles = {}
+    @game_teams.each do |game_team|
+      if game_ids_for_season.include?(game_team.game_id)
+        team_tackles[game_team.team_id] ||= 0
+        team_tackles[game_team.team_id] += game_team.tackles
+      end
+    end
+    
+    team_with_least_tackles = nil
+    least_tackles = 1000000 #random number I chose, just needed a big enough number
```
Comment (DustinPeukert) [inline lib/stat_calculator.rb:224]: Same here:
- may be able to accomplish this method with a min_by in this block

Out of curiosity, what is the purpose of setting least_tackles to 100000?
Instead, could setting it to the first teams number of tackles work?

**Review by DustinPeukert — APPROVED** [review]
I left some comments for clarification and suggestions. As long as these work for now and pass all tests then I think it's fine to merge. Just keep these in mind for possible refactoring opportunities!

File: lib/stat_calculator.rb (hunk 2 above)
Comment (kevin-newland) [reply lib/stat_calculator.rb:224]: The purpose of setting it to a high number was to ensure that teams tackle count will be smaller. 
I do realize there is room for refactoring. I was under the impression we were trying to get working code first. This is the just way my brain was able to put all the pieces of the puzzle together.
If you choose not to merge it I undersstand, ill push changes up tomorrow.

File: lib/stat_calculator.rb (hunk 1 above)
Comment (kevin-newland) [reply lib/stat_calculator.rb:198]: I do agree that it can be refactored. This was just how my brain was able to put it all together.

File: lib/stat_calculator.rb (hunk 2 above)
Comment (DustinPeukert) [reply lib/stat_calculator.rb:224]: Nope, it's all good. Like I said in my main review comment. I'm ok with merging this. Just wanted to spark some refactor ideas for you for later!

Comment (kevin-newland) [reply lib/stat_calculator.rb:224]: I appreciate the advice and the review!

File: lib/stat_calculator.rb [hunk 3]
```diff
@@ -171,4 +171,68 @@ def worst_offense
     worst_offense = @teams.find { |team| team.team_id.to_s == worst_offense_team.to_s}
     worst_offense.team_name
   end
+
```
Comment (qoryhanisagal) [inline lib/stat_calculator.rb:174]: Suggestions: Both methods iterate through @teams to find the matching team after determining the team ID.
Solution: Use a hash (team_lookup) to map team_id to team_name for constant-time lookups.

```ruby
team_lookup = @teams.each_with_object({}) do |team, hash|
  hash[team.team_id.to_s] = team.team_name
end
```
**Then, replace the final loop with:**

```ruby
team_lookup[team_with_most_tackles.to_s]
```

File: lib/stat_calculator.rb [hunk 4]
```diff
@@ -171,4 +171,68 @@ def worst_offense
...
+   
+    team_tackles = {}
+    @game_teams.each do |game_team|
+      if game_ids_for_season.include?(game_team.game_id)
+        team_tackles[game_team.team_id] ||= 0
+        team_tackles[game_team.team_id] += game_team.tackles
+      end
+    end
+    
+    team_with_least_tackles = nil
```
Comment (qoryhanisagal) [inline lib/stat_calculator.rb:223]: I know this probably due to the spec harness values

Issue: The least_tackles variable is initialized to 1000000, which is arbitrary and may cause confusion.
Solution: Initialize least_tackles to Float::INFINITY to avoid magic numbers.

File: lib/stat_calculator.rb
Comment (qoryhanisagal) [inline lib/stat_calculator.rb:1]: Overall look great Kevin
I offer this suggestion regarding repetition in code.
- **Issue:** Both methods share nearly identical code for filtering games and aggregating tackles.
- **Solution:** Extract the shared logic into a helper method that returns the team_tackles hash for a given season.

```ruby

def tackles_by_team_for_season(season)
  game_ids_for_season = @games.select { |game| game.season == season }.map(&:game_id)
  team_tackles = Hash.new(0)

  @game_teams.each do |game_team|
    team_tackles[game_team.team_id] += game_team.tackles if game_ids_for_season.include?(game_team.game_id)
  end

  team_tackles
end
```

Then, reuse this in `most_tackles`and `fewest_tackles`.
**Test Coverage**: Verifying that ties are handled correctly like multiple teams with the same tackles

Refactored base on suggestions **I used Pseudocode for this so please double check**
```ruby
def tackles_by_team_for_season(season)
  game_ids_for_season = @games.select { |game| game.season == season }.map(&:game_id)
  team_tackles = Hash.new(0)

  @game_teams.each do |game_team|
    team_tackles[game_team.team_id] += game_team.tackles if game_ids_for_season.include?(game_team.game_id)
  end

  team_tackles
end

def most_tackles(season)
  team_tackles = tackles_by_team_for_season(season)
  team_with_most_tackles = team_tackles.max_by { |_, tackles| tackles }.first

  @teams.find { |team| team.team_id.to_s == team_with_most_tackles.to_s }.team_name
end

def fewest_tackles(season)
  team_tackles = tackles_by_team_for_season(season)
  team_with_least_tackles = team_tackles.min_by { |_, tackles| tackles }.first

  @teams.find { |team| team.team_id.to_s == team_with_least_tackles.to_s }.team_name
end
```

File: spec/stat_calculator_spec.rb [hunk 5]
```diff
@@ -90,14 +90,28 @@
...
     end
   end
 
   describe '#worst_offense' do
-    it "Returns name of the team with the lowest average number of goals scored per game across all seasons" do
+    it "returns name of the team with the lowest average number of goals scored per game across all seasons" do
       expect(@stat_calculator.worst_offense).to eq("Utah Royals FC")
     end
   end
+
```
Comment (qoryhanisagal) [inline spec/stat_calculator_spec.rb:103]: **Update Tests**
The existing tests for most_tackles and fewest_tackles are sufficient, but you could add edge cases:
```ruby
describe '#most_tackles' do
  it 'returns the name of the team with the most tackles for a specific season' do
    expect(@stat_calculator.most_tackles("20132014")).to eq("FC Cincinnati")
    expect(@stat_calculator.most_tackles("20142015")).to eq("Seattle Sounders FC")
  end

  it 'returns nil if there are no games for the season' do
    expect(@stat_calculator.most_tackles("99999999")).to be_nil
  end
end

describe '#fewest_tackles' do
  it 'returns the name of the team with the fewest tackles for a specific season' do
    expect(@stat_calculator.fewest_tackles("20132014")).to eq("Atlanta United")
    expect(@stat_calculator.fewest_tackles("20142015")).to eq("Orlando City SC")
  end

  it 'returns nil if there are no games for the season' do
    expect(@stat_calculator.fewest_tackles("99999999")).to be_nil
  end
end
```

**Review by qoryhanisagal — APPROVED** [review]
@kevin-newland your code demonstrate "you understood the assignment"  😂 Splendid Job 😊!!! I can tell in your code you had a clear approach to solving it. While it can be optimized and made more modular, the logic is correct and adheres to the prompt’s requirements. The key positive aspects are its clarity, logical flow, and correct use of data structures for dynamic handling of tackles by season.

**Positives in This Code**

**1.	Logical Step-by-Step Approach:**
- The code breaks down the problem clearly:
	1.	Filter games for the specified season.
	2.	Aggregate tackles for teams participating in those games.
	3.	Identify the team with the fewest tackles.
	4.	Retrieve and return the team name.

**2.	Dynamic Handling of Season Data:**
- The use of `game_ids_for_season` ensures that the code dynamically handles games for any given season, rather than hardcoding any values.

**3.	Proper Use of a Hash:**
- The `team_tackles` hash is initialized with an appropriate pattern for accumulating values:
```ruby
team_tackles[game_team.team_id] ||= 0
team_tackles[game_team.team_id] += game_team.tackles
```
**4.	Focus on Tackles for Relevant Teams:**
- By filtering `game_team` data using `game_ids_for_season`, the code ensures that only relevant games for the season are considered.

**5.	Readable Code:**
- Variable names like `game_ids_for_season`,`team_tackles`, and `team_with_least_tackles` make the code easier to follow and self-explanatory.

**6.	Straightforward Identification of the Team:**
- The use of the final loop to match team_with_least_tackles with its corresponding team_name is simple and effective.

**Room for Improvement**

@kevin-newland  this was extremely well thoughtout code very effective and readable, it could benefit from:
**1.	Efficiency:**
- The nested loops (@games, @game_teams, @teams) result in a higher time complexity. Using a single pass or helper methods could make it more efficient.

**2.	Modularity:**
- The method could be split into smaller, reusable helper methods for filtering games, aggregating tackles, and finding the team.
- 
**3.	Error Handling:**
- The code doesn’t account for edge cases which isnt a requirement, however if you were consider asking yourself:
- What happens if no games are found for the given season?
- What happens if @teams is empty or doesn’t contain a matching team_id?

**4.	Initialization of least_tackles:**
- The variable `least_tackles` is not initialized in the provided code, which would cause an error. A large initial value (e.g., Float::INFINITY) could look like this:
```ruby
least_tackles = Float::INFINITY
```
	
**Tip: I dropped the refactored code in the comments**

**Review by KMPWeiland — APPROVED** [review]
Nice job tackling these gnarly methods! Agreed -- working code is best code! We can discuss refactor opps during our Monday call.  Well done :)
